app/src/main/java/org/dhis2/uicomponents/map/carousel/CarouselProgramEventHolder.kt

```diff
@@ -5,59 +5,128 @@ import android.text.SpannableString
-                    itemView.context.getString(R.string.event_event)
-                        .toLowerCase(Locale.getDefault())
-                )
+    private fun hideEventValueLayout() {
+        binding.showValuesButton.visibility = View.INVISIBLE
+        binding.dataElementListGuideline.visibility = View.INVISIBLE
+        binding.dataElementList.visibility = View.GONE
+        binding.dataValue.text = itemView.context.getString(R.string.no_data)
+        binding.showValuesButton.setOnClickListener(null)
+    }
+
+    private fun initValues(
+        valueListIsOpen: Boolean,
+        dataElementValues: MutableList<Pair<String, String>>
+    ) {
+        binding.dataElementList.removeAllViews()
+        binding.dataValue.text = null
+        binding.showValuesButton.scaleY = if (valueListIsOpen) 1f else -1f
+        binding.showValuesButton
+            .animate()
+            .scaleY(if (valueListIsOpen) -1f else 1f)
+            .setDuration(500)
+            .withStartAction { binding.showValuesButton.scaleY = if (valueListIsOpen) 1f else -1f }
+            .withEndAction { binding.showValuesButton.scaleY = if (valueListIsOpen) -1f else 1f }
+            .start()
+        if (valueListIsOpen) {
+            binding.dataElementListGuideline.visibility = View.VISIBLE
+            binding.dataElementList.visibility = View.VISIBLE
+            for (nameValuePair in dataElementValues) {
+                val fieldValueBinding: ItemFieldValueBinding =
+                    ItemFieldValueBinding.inflate(
+                        LayoutInflater.from(binding.dataElementList.context)
+                    )
+                fieldValueBinding.name = nameValuePair.val0()
+                fieldValueBinding.value = nameValuePair.val1()
+                binding.dataElementList.addView(fieldValueBinding.root)
+            }
         } else {
-            binding.noCoordinatesLabel.root.visibility = View.INVISIBLE
+            binding.dataElementListGuideline.visibility = View.INVISIBLE
+            binding.dataElementList.visibility = View.GONE
+            val stringBuilder =
+                SpannableStringBuilder()
+            for (nameValuePair in dataElementValues) {
+                if (nameValuePair.val1() != "-") {
+                    val value =
+                        SpannableString(nameValuePair.val1())
+                    val colorToUse =
+                        if (dataElementValues.indexOf(nameValuePair) % 2 == 0) {
+                            Color.parseColor("#8A333333")
+                        } else {
+                            Color.parseColor("#61333333")
+                        }
+                    value.setSpan(
+                        ForegroundColorSpan(colorToUse),
+                        0,
+                        value.length,
+                        Spanned.SPAN_INCLUSIVE_INCLUSIVE
+                    )
+                    stringBuilder.append(value)
```

Comment:
Do you think it makes sense to extract this presentation logic? This is a very small example but when it gets bigger we cannot unit test :(

Reply:
Okey